### DISCUSSION TITLE
Conveyor 15.0

### DISCUSSION THREAD

**Discussion by mikehearn** [hydraulic-software/conveyor#121] [Announcements]
This is a feature release.

# [Update control API](https://conveyor.hydraulic.dev/15.0/release-notes/#update-control-api)

Conveyor 15 adds an API that your app can use to interact with online updates. It's available for [Electron](https://conveyor.hydraulic.dev/15.0/control-api-electron/), [JVM](https://conveyor.hydraulic.dev/15.0/control-api-jvm/) and [native](https://conveyor.hydraulic.dev/15.0/control-api-native/) apps. It lets you check if there are new versions available at the update site as well as trigger an update/restart cycle, and so can be used to implement alternative update policies that aren't supported out of the box.

The [app.updates](https://conveyor.hydraulic.dev/15.0/configs/update-modes/) key now supports MANUAL as well as NONE. The latter lets you disable update support entirely, yielding a smaller package in some cases. It's appropriate when you don't want Conveyor to handle updates for you. MANUAL turns off automatic updates whilst still allowing updates to be triggered using the control API.

The scaffold apps you can make using conveyor generate have been updated to demonstrate how to use this API.

# [Windows](https://conveyor.hydraulic.dev/15.0/release-notes/#windows)

[The app.windows.package-extras key and CPU specific variants](https://conveyor.hydraulic.dev/15.0/configs/windows/#appwindowsamd64aarch64package-extras) allow files to be added outside the app-specific subdirectory. This is only useful for Electron and JVM apps where the root inputs are relocated to a conventional subdirectory that keeps them separated from the runtime.

When signing with Sectigo certificates Conveyor now forces the use of the legacy cross-signed root. This improves compatibility with Windows 10 machines that aren't properly downloading root store updates from Microsoft, which can happen due to incorrect or over-aggressive operating system settings.

# [macOS](https://conveyor.hydraulic.dev/15.0/release-notes/#macos)

[The app.mac.skip-framework-symlink-removal key](https://conveyor.hydraulic.dev/15.0/configs/mac/) allows frameworks to be exempted from the simplifying transform that removes redundant symlinks from frameworks.

You can now configure update checks to occur as frequently as once a minute, down from the previous limit of once an hour. This makes testing updates integration easier.

# [Electron](https://conveyor.hydraulic.dev/15.0/release-notes/#electron)

The handling of symlinked packages has been improved. This is useful when your app depends on NPM packages that aren't uploaded to a package registry and where you've used npm's symlink ability to point into your dev tree.

**Comment by ennerf** [comment]
@mikehearn Thanks for releasing the control API! 

I saw that the [docs](https://conveyor.hydraulic.dev/15.0/control-api-jvm/) mention that the API is not supported on Linux. Is it  completely unavailable or is it just unable to update the app by itself?

**Reply by mikehearn** [reply]
Getting the version number from `metadata.properties` is supported, triggering the APT GUI and doing a restart isn't. To implement it we'd need to figure out what the right command is to pop up the sudo window, trigger an update, then monitor the filing system and wait until the package has finished installing (some postinst script that sets a flag or something), and then do the restart. It's not that hard to do, it's just that most Linux users prefer to control updates centrally through their package manager so it didn't make the cut this time.

**Reply by ennerf** [reply]
Thanks for clarifying. I just need to show users a popup where they can copy the appropriate update command and copy & paste it into a terminal themselves.
